onadata/libs/utils/viewer_tools.py

```diff
@@ -340,10 +343,17 @@ def get_enketo_edit_url(request, instance, return_url):
     return url
 
 
-def get_enketo_preview_url(request, username, id_string, xform_pk=None):
+def get_enketo_preview_url(
+        request, id_string, username=None, xform_pk=None, xform_uuid=None):
     """Return an Enketo preview URL."""
-    form_url = get_form_url(
-        request, username, settings.ENKETO_PROTOCOL, True, xform_pk=xform_pk)
+    if xform_pk and username:
```

Comment:
if we're adding the keys to kwargs below, we should definitely add them here too, eg say `username=username` etc.

But, I wonder if we should define kwargs all at once and have a single call to `get_form_url` like

```python
get_form_url_kwargs = {
    "protocol": settings.ENKETO_PROTOCOL,
    "preview": True
}
if xform_pk and username:
    get_form_url_kwargs.update{
        "username": username,
        "xform_pk": xform_pk
    }
elif xform_uuid:
    get_form_url_kwargs.update{
        "xform_uuid": xform_uuid
    }

form_url = get_form_url(request, **get_form_url_kwargs)
```

Reply:
I'm not sold on defining the kwargs actually being better, thoughts?

Reply:
I think this would work fine. There are no downsides of using an array here, would help reduce the duplicate calls to `get_form_url` changing this to:

```
    form_url_kwargs = {
        "request": request,
        "protocol": settings.ENKETO_PROTOCOL,
        "preview": True
    }

    if xform_pk and username:
        form_url_kwargs.update({
            'xform_pk': xform_pk,
            'username': username
        })
    elif xform_uuid:
        form_url_kwargs.update({
            'xform_uuid': xform_uuid
        })

    form_url = get_form_url(**form_url_kwargs)
    values = {'form_id': id_string, 'server_url': form_url}
```

Reply:
lgtm